File: problems/p0347_top_k_frequent_elements/solution-0347.java
package p0347_top_k_frequent_elements;

import java.util.Arrays;
import java.util.HashMap;
import java.util.PriorityQueue;
import java.util.Random;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import org.junit.jupiter.api.Test;

class Solution {

    private int[] unique;
    private HashMap<Integer, Integer> count = new HashMap<>();

    public int[] topKFrequent(int[] nums, int k) {
        if (k == 0) {
            return new int[]{};
        }

        count = new HashMap<>();
        for (int num : nums) {
            count.put(num, count.getOrDefault(num, 0) + 1);
        }

        PriorityQueue<int[]> heap = new PriorityQueue<>((a, b) -> {
            // Difference in frequency
            return b[1] - a[1];
        });

        for (HashMap.Entry<Integer, Integer> entry : count.entrySet()) {
            heap.add(
                    new int[]{(int) entry.getKey(), (int) entry.getValue()}
            );
        }

        // Gets the k most frequent elements
        int[] result = new int[k];
        for (int i = 0; i < k; i++) {
            result[i] = heap.poll()[0];
        }
        return result;
    }

    int frequency(int index) {
        int num = unique[index];
        return count.get(num);
    }

    int swap(int i, int j) {
        int temp = unique[i];
        unique[i] = unique[j];
        unique[j] = temp;
        return 0;
    }

    int partition(int left, int right, int pivotIndex) {
        int pivotFreq = frequency(pivotIndex);

        swap(pivotIndex, right);
        int storeIndex = left;
        for (int i = left; i <= right; i++) {
            int iFreq = frequency(i);
            if (iFreq > pivotFreq) {
                swap(storeIndex, i);
                storeIndex++;
            }
        }

        swap(right, storeIndex);
        return storeIndex;
    }

    void quickSelect(int left, int right, int k) {
        if (left == right) {
            return;
        }

        Random rand = new Random();
        int pivotIndex = left + rand.nextInt(right - left + 1);
        pivotIndex = partition(left, right, pivotIndex);
        if (k == pivotIndex) {
            // Found k most frequent elements
        } else if (k < pivotIndex) {
            // Go left, too many elements
            quickSelect(left, pivotIndex - 1, k);
        } else {
            // Go right, not enough elements
            quickSelect(pivotIndex + 1, right, k);
        }
    }

    int[] topKFrequentQuickSelect(int[] nums, int k) {
        if (k == 0) {
            return new int[]{};
        }
        count = new HashMap<>();
        for (int num : nums) {
            count.put(num, count.getOrDefault(num, 0) + 1);
        }

        unique = new int[count.size()];
        int index = 0;
        for (int num : count.keySet()) {
            unique[index++] = num;
        }

        quickSelect(0, unique.length - 1, k - 1);

        int[] result = new int[k];
        System.arraycopy(unique, 0, result, 0, k);
        return result;
    }
}

class TestCase {

    int[] nums;
    int k;
    int[] expected;

    TestCase(int[] nums, int k, int[] expected) {
        this.nums = nums;
        this.k = k;
        this.expected = expected;
    }
}

@SuppressWarnings("unused")
class SolutionTest {

    @Test
    void testTopKFrequent() {
        Solution solution = new Solution();
        TestCase[] testCases = new TestCase[]{
            new TestCase(new int[]{1, 1, 1, 2, 2, 3}, 2, new int[]{1, 2}),
            new TestCase(new int[]{1}, 1, new int[]{1}),
            new TestCase(new int[]{4, 1, -1, 2, -1, 2, 3}, 2, new int[]{-1, 2}),};
        for (TestCase testCase : testCases) {
            System.err.println("Input: " + java.util.Arrays.toString(testCase.nums) + " k: " + testCase.k);
            System.err.println("Expected: " + java.util.Arrays.toString(testCase.expected));
            int[] result = solution.topKFrequent(testCase.nums, testCase.k);
            System.err.println("Result: " + java.util.Arrays.toString(result));

            Arrays.sort(result);
            Arrays.sort(testCase.expected);
            assertArrayEquals(testCase.expected, result);

            int[] resultQS = solution.topKFrequentQuickSelect(testCase.nums, testCase.k);
            System.err.println("Result (QS): " + java.util.Arrays.toString(resultQS));
            Arrays.sort(resultQS);
            assertArrayEquals(testCase.expected, resultQS);
        }
    }
}

File: problems/p0347_top_k_frequent_elements/solution-0347.cpp
#include <vector>
#include <unordered_map>
#include <cstdint>
#include <chrono>
#include <algorithm>
#include <random>
#include <cassert>
#include <fmt/core.h>
#include <fmt/ranges.h>

class Solution {
private:
    std::unordered_map<int, uint32_t> count;
    std::vector<int> unique;
public:

    uint32_t random_in_range_incl(uint32_t start, uint32_t end) {
        assert(end - start + 1 != 0);
        return start + rand() % (end - start + 1);
    }

    uint32_t partition(uint32_t left, uint32_t right, uint32_t pivot_index) {
        uint32_t pivot_freq = count[unique[pivot_index]];
        std::swap(unique[pivot_index], unique[right]);

        uint32_t store_index = left;
        for (uint32_t i = left; i <= right; ++i) {
            uint32_t i_freq = count[unique[i]];
            if (i_freq > pivot_freq) {
                std::swap(unique[store_index], unique[i]);
                store_index++;
            }
        }

        std::swap(unique[right], unique[store_index]);

        return store_index;
    }

    void quick_select(uint32_t left, uint32_t right, uint32_t k) {
        if (left >= right) return;

        uint32_t pivot_index = random_in_range_incl(left, right);
        pivot_index = partition(left, right, pivot_index);

        if (k == pivot_index) {
            return;
        } else if (k < pivot_index) {
            quick_select(left, pivot_index - 1, k);
        } else {
            quick_select(pivot_index + 1, right, k);
        }
    }
    std::vector<int> topKFrequent(std::vector<int>& nums, int k) {
        if (k == 0)  return std::vector<int>();
        uint32_t k_ = k;
        count = std::unordered_map<int, uint32_t>();
        unique = std::vector<int>();

        for (int num : nums) {
            if (count.contains(num)) {
                count[num]++;
            } else {
                count[num] = 1;
                unique.push_back(num);
            }
        }

        uint32_t unique_nb = unique.size();

        if (k_ == unique_nb) return unique;

        quick_select(0, unique_nb - 1, k_);
        return std::vector<int>(unique.begin(), unique.begin() + k);
    }
};

struct TestCase {
    std::vector<int> nums;
    int k;
    std::vector<int> expected;
};

int test() {
    Solution solution;
    TestCase test_cases[] = {
        { {1,1,1,2,2,3}, 2, {1,2} },
        { {1}, 1, {1} },
        { {4,4,4,6,6,6,6,7,7,8}, 3, {6,4,7} },
        { {4,1,-1,2,-1,2,3}, 2, {-1,2} }
    };
    int error_found = 0;
    for (auto& test_case : test_cases) {
        std::vector<int> result = solution.topKFrequent(test_case.nums, test_case.k);
        std::sort(result.begin(), result.end());
        std::sort(test_case.expected.begin(), test_case.expected.end());
        if (result != test_case.expected) {
            fmt::println(stderr,
                "Test failed for input nums: {}, k: {}. Expected output: {} but got: {}",
                fmt::join(test_case.nums, ", "),
                test_case.k,
                fmt::join(test_case.expected, ", "),
                fmt::join(result, ", ")
            );
            ++error_found;
        }
    }
    return error_found;
}

int main([[maybe_unused]] int argc, [[maybe_unused]] char* argv[]) {
    int errors = test();
    if (errors == 0) {
        fmt::println(stderr, "All test cases passed!");
    } else {
        fmt::println(stderr, "{} test case(s) failed.", errors);
        return 1;
    }
    return 0;
}
